ir/src/main/scala/org/scalajs/ir/ScalaJSVersions.scala

```diff
@@ -12,32 +12,94 @@
 
 package org.scalajs.ir
 
-object ScalaJSVersions {
+import java.util.concurrent.ConcurrentHashMap
 
-  /* DO NOT MAKE THESE 'final val's!
-   * When referring to these "constants" from separate libraries, if it is a
-   * 'final val', the value will be copied in the binaries of those libraries.
-   * If they are then linked to a different version of the IR artifact, their
-   * copy of these constants will not be updated.
-   */
+import scala.util.matching.Regex
 
+object ScalaJSVersions extends VersionChecks(
   /** Scala.js version. */
-  val current: String = "1.0.0-SNAPSHOT"
-
-  /** Version of binary IR emitted by this version of Scala.js.
-   *
-   *  This should be either of:
-   *  - a prior release version (i.e. "0.5.0", *not* "0.5.0-SNAPSHOT")
-   *  - `current`
-   */
-  val binaryEmitted: String = current
-
-  /** Versions whose binary files we can support (used by deserializer) */
-  val binarySupported: Set[String] = {
-    Set(binaryEmitted)
+  current = "1.0.0-SNAPSHOT",
+  /** Version of binary IR emitted by this version of Scala.js. */
+  binaryEmitted = "1.0-SNAPSHOT"
+)
+
+/** Helper class to allow for testing of logic. */
+class VersionChecks private[ir] (
+    final val current: String,
+    final val binaryEmitted: String
+) {
+  import VersionChecks._
+
+  checkConsistent(current, binaryEmitted)
+
+  private val (binaryMajor, binaryMinor, binaryPreRelease) = parseBinary(binaryEmitted)
+
+  private val knownSupportedBinary = {
+    val m = new ConcurrentHashMap[String, Unit]()
+    m.put(binaryEmitted, ())
+    m
+  }
+
+  /** Check we can support this binary version (used by deserializer) */
+  final def checkSupported(version: String): Unit = {
+    if (!knownSupportedBinary.containsKey(version)) {
+      val (major, minor, preRelease) = parseBinary(version)
+      val supported = (
+          // the exact pre-release version is supported via knownSupportedBinary
+          preRelease.isEmpty &&
+          major == binaryMajor &&
+          minor <= binaryMinor &&
+          (binaryPreRelease.isEmpty || minor < binaryMinor)
+      )
+
+      if (supported) {
+        knownSupportedBinary.put(version, ())
+      } else {
+        throw new IRVersionNotSupportedException(version, binaryEmitted,
+            s"This version ($version) of Scala.js IR is not supported. " +
+            s"Supported versions are up to $binaryEmitted")
+      }
+    }
   }
+}
 
-  // Just to be extra safe
-  assert(binarySupported contains binaryEmitted)
+private object VersionChecks {
+  private val binaryRE = """^([0-9]+)\.([0-9]+)(-.*)?$""".r
+
+  private def parseBinary(v: String): (Int, Int, Option[String]) = {
+    val m = mustMatch(binaryRE, v)
+    (m.group(1).toInt, m.group(2).toInt, Option(m.group(3)))
```

Comment:
```suggestion
    (m.group(1).toInt, m.group(2).toInt, Option(m.group(3)).map(_.substring(1)))
```
? So that the `-` does not belong to the result string.